src/consensus/blockruleindex.cpp

```diff
@@ -0,0 +1,300 @@
+    if (pstart && m_ruleStateMap.count(pstart))
+        prevRuleStates = m_ruleStateMap[pstart];
+
+    // Assign the rule states for the new block
+    //   1) Set all assigned rules for new block to DEFINED (we'll check whether they are active next)
+    std::set<int> assignedRules = m_deployments->GetAssignedRules(currentMedianTime);
+    for (std::set<int>::iterator it = assignedRules.begin(); it != assignedRules.end(); ++it)
+    {
+        newRuleStates[*it] = DEFINED;
+    }
+
+    //   2) Add all failed and active rules
+    std::set<int> failedRules;
+    std::set<int> activeRules;
+    for (RuleStates::iterator it = prevRuleStates.begin(); it != prevRuleStates.end(); ++it)
+    {
+        switch (it->second)
+        {
+        case FAILED:
+            newRuleStates[it->first] = FAILED;
+            failedRules.insert(it->first);
+            break;
+
+        case LOCKED_IN:
+            // Locked in rules become active at this block
+
+        case ACTIVE:
+            newRuleStates[it->first] = ACTIVE;
+            activeRules.insert(it->first);
+            break;
+
+        default:
+            break;
+        }
+    }
+
+    //   3) Count set bits for interval
+    typedef std::map<int /*rule*/, int /*count*/> RuleCountMap;
+    RuleCountMap ruleCountMap;
+    while (pprev)
+    {
+        if (UsesVersionBits(pprev->nVersion))
+        {
+            for (int b = MIN_BIT; b <= MAX_BIT; b++)
+            {
+                if ((pprev->nVersion >> b) & 0x1)
+                {
+                    int rule = m_deployments->GetAssignedRule(b, pprev->GetMedianTimePast());
+
+                    // Only count bits for soft forks that are assigned and are not failed and not already active
+                    if ((rule == NO_RULE) || failedRules.count(rule) || activeRules.count(rule))
+                        continue;
+
+                    if (ruleCountMap.count(rule))
+                        ruleCountMap[rule]++;
+                    else
+                        ruleCountMap[rule] = 1;
+                }
+            }
+        }
+
+        if (pprev == pstart)
+            break;
+
+        pprev = pprev->pprev;
+    }
+
+    //   4) Set newly locked in and failed states
+    RuleCountMap::iterator it = ruleCountMap.begin();
+    for (; it != ruleCountMap.end(); ++it)
+    {
+        int rule = it->first;
+        int count = it->second;
+        const SoftFork* softFork = m_deployments->GetSoftFork(rule);
+
+        if (softFork && (count >= softFork->GetThreshold()))
+            newRuleStates[rule] = LOCKED_IN;
+        else
+            if (!m_deployments->IsRuleAssigned(rule, currentMedianTime))
+                newRuleStates[rule] = FAILED;
+    }
+
+    m_ruleStateMap[pblockIndex] = newRuleStates;
+
+    return true;
+}
+
+void BlockRuleIndex::Clear()
+{
+    m_ruleStateMap.clear();
+}
+
+
+static BlockRuleIndex g_blockRuleIndex;
+BlockRuleIndex& Consensus::VersionBits::GetBlockRuleIndex() { return g_blockRuleIndex; }
```

Comment:
Why the getter for the new global? If you are introducing a new global that consensus code will rely on, why not use the global as explicitly a possible?
Just declare g_blockRuleIndex as extern in the .h file, s/Consensus::VersionBits::GetBlockRuleIndex()/g_blockRuleIndex and simply get rid of Consensus::VersionBits::GetBlockRuleIndex().